src/components/PageMetadataEdit.tsx

```diff
@@ -0,0 +1,1418 @@
+    url: string | undefined;
+}
+const FormGroupContactsReportIssue: FunctionComponent<
+    FormGroupContactsReportIssueProps
+> = (props) => {
+    const url = _.defaultTo(props.url, '');
+    const dispatch = useContext(MetadataDispatchContext);
+    useEffect(() => doValidation(url), []);
+    const [validated, setValidated] = useState<Validate>('error');
+    if (_.isNull(dispatch)) return null;
+    const doValidation = (s: string) => {
+        if (s === '') {
+            setValidated('default');
+        } else if (validator.isURL(s)) {
+            setValidated('success');
+        } else {
+            setValidated('error');
+        }
+    };
+    const onChange = (value: string) => {
+        dispatch({ type: 'contactsReportIssue', url: value });
+        doValidation(value);
+    };
+    const element = (
+        <FormGroup
+            label="Report issue URL"
+            fieldId="form-contacts-report-issue"
+            helperText="URL format"
+            helperTextInvalid="Invalid url"
+            helperTextInvalidIcon={<ExclamationCircleIcon />}
+            validated={validated}
+        >
+            <TextInput
+                isRequired
+                type="text"
+                placeholder="https://url.corp.redhat.com/rut-report-issue"
+                id="form-contacts-report-issue"
+                name="contacts-report-issue"
+                onChange={onChange}
+                validated={validated}
+                value={url}
+            />
+        </FormGroup>
+    );
+    return element;
+};
+
+/*
+ * Next technique based on recomendation:
+ * https://reactjs.org/docs/hooks-faq.html#how-to-avoid-passing-callbacks-down
+ */
+const MetadataDispatchContext =
+    createContext<React.Dispatch<MetadataAction> | null>(null);
+
+type MetadataAction =
+    | { type: 'metadataLoaded'; metadata: MetadataRaw }
+    | { type: 'newIssue' }
+    | { type: 'rmIssue'; issue: KnownIssue }
+    | { type: 'updateIssue'; issue: KnownIssue; updatedIssue: KnownIssue }
+    | { type: 'newDependency' }
+    | { type: 'rmDependency'; dependency: Dependency }
+    | {
+          type: 'updateDependency';
+          dependency: Dependency;
+          updatedDependency: Dependency;
+      }
+    | { type: 'productVersion'; productVersion: string }
+    | { type: 'description'; description?: string }
+    | { type: 'testcaseName'; testcaseName: string }
+    | { type: 'testcaseNameIsRegex'; isRegex: boolean }
+    | { type: 'priority'; priority: number }
+    | { type: 'contactsCIName'; name?: string }
+    | { type: 'contactsTeamName'; team?: string }
+    | { type: 'contactsDocs'; docs?: string }
+    | { type: 'contactsCiSystemURL'; url?: string }
+    | { type: 'contactsEmail'; email?: string }
+    | { type: 'contactsGchat'; gchat?: string }
+    | { type: 'contactsIrc'; irc?: string }
+    | { type: 'contactsReportIssue'; url?: string };
+
+const metadataReducer: Reducer<MetadataRaw, MetadataAction> = (
```

Comment:
I'm thinking this should be in the `src/reducers/` directory, right?

Reply:
redux != react reducers.
@mgrabovsky : again: "I'm thinking" doesn't help in review process.

Reply:
I don't understand your comment. What is the difference between a reducer and a reducer? Both are just pure functions that take an action and old state and return a new state. This piece of code is just processing events and modifying state. It has nothing to do with the UI and by the principle of separation of concerns, it should be in a separate module.

Reply:
1. One reducer is site wide.
2. Seconde reducer is component oriented. Only 1 component uses it.

In `src/reducers/` there is redux-site wide reducer. They hold state for all side. Mixing to mechanism, and putting reducer for component is bad. Answer: no , component reducer should not be in `src/reducers/`

Reply:
Find a way for a "good" aprioach for splitting component React-native reducers into separate file. Do some research next time.